### PR TITLE
facts: always set osd_pool_default_crush_rule

### PR DESCRIPTION
**Pull request by guits**
since `osd_pool_default_crush_rule` is used as a default value for the
rule name when it's not specified by the user, this fact always needs to
be set in any cases.

Closes: https://bugzilla.redhat.com/show_bug.cgi?id=1817586

Signed-off-by: Guillaume Abrioux <gabrioux@redhat.com>

### CODE REVIEW

**Comment by guits** [comment]
jenkins test centos-container-update

**Comment by dsavineau** [comment]
jenkins test centos-container-all_daemons

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-non_container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-external_clients

**Comment by guits** [comment]
jenkins test centos-container-podman

**Comment by guits** [comment]
jenkins test centos-non_container-lvm_osds

**Comment by guits** [comment]
jenkins test centos-container-ooo_collocation

**Comment by guits** [comment]
jenkins test centos-non_container-update

**Comment by guits** [comment]
jenkins test centos-non_container-all_daemons

**Comment by guits** [comment]
jenkins test centos-container-podman

File: roles/ceph-client/tasks/create_users_keys.yml [hunk 1]
```diff
@@ -69,6 +69,11 @@
     - condition_copy_admin_key | bool
     - inventory_hostname == groups.get('_filtered_clients', []) | first
   block:
+    - import_role:
+        name: ceph-facts
+        tasks_from: get_def_crush_rule_name.yml
+      when: groups.get(mon_group_name, []) | length > 0
```
Comment (guits) [inline roles/ceph-client/tasks/create_users_keys.yml]:
```suggestion
      when: groups.get(mon_group_name, []) | length > 0 or condition_copy_admin_key | bool
```

File: roles/ceph-client/tasks/create_users_keys.yml [hunk 2]
```diff
@@ -78,15 +83,23 @@
       failed_when: false
       delegate_to: "{{ delegated_node }}"
 
+    - name: set_fact container_exec_cmd
+      set_fact:
+        container_run_cmd: "{{ container_binary + ' run --rm --net=host -v /etc/ceph:/etc/ceph:z -v /var/lib/ceph/:/var/lib/ceph/:z -v /var/log/ceph/:/var/log/ceph/:z --entrypoint=ceph ' + ceph_docker_registry + '/' + ceph_docker_image + ':' + ceph_docker_image_tag if containerized_deployment else '' }}"
```
Comment (guits) [inline roles/ceph-client/tasks/create_users_keys.yml]:
```suggestion
        container_run_cmd: "{{ container_binary + ' run --rm --net=host -v /etc/ceph:/etc/ceph:z -v /var/lib/ceph/:/var/lib/ceph/:z -v /var/log/ceph/:/var/log/ceph/:z --entrypoint=ceph ' + ceph_docker_registry + '/' + ceph_docker_image + ':' + ceph_docker_image_tag if containerized_deployment else 'ceph' + ' -n client.admin -k /etc/ceph/' + cluster + '.client.admin.keyring' }}"
```

File: roles/ceph-client/tasks/create_users_keys.yml [hunk 3]
```diff
@@ -78,15 +83,23 @@
...
+
+    - name: set_fact container_exec_cmd
+      set_fact:
+        container_exec_cmd: "{{ container_run_cmd if containerized_deployment | bool else 'ceph' + ' -n client.admin -k /etc/ceph/' + cluster + '.client.admin.keyring' }}"
```
Comment (guits) [inline roles/ceph-client/tasks/create_users_keys.yml]:
```suggestion
```

File: roles/ceph-facts/tasks/get_def_crush_rule_name.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+---
+- name: get current default crush rule details
+  command: "{{ container_exec_cmd | default('') }} ceph --cluster {{ cluster }} osd -f json crush rule dump"
```
Comment (guits) [inline roles/ceph-facts/tasks/get_def_crush_rule_name.yml]:
```suggestion
  command: "{{ container_exec_cmd | default('') }} ceph --cluster {{ cluster }} -f json osd crush rule dump"
```

Comment (guits) [inline roles/ceph-facts/tasks/get_def_crush_rule_name.yml]:
```suggestion
  command: "{{ hostvars[delegated_node | default(groups[mon_group_name][0])]['container_exec_cmd'] }} ceph --cluster {{ cluster }} osd -f json crush rule dump"
```

File: roles/ceph-client/tasks/create_users_keys.yml (hunk 1 above)
Comment (guits) [inline roles/ceph-client/tasks/create_users_keys.yml]:
```suggestion
```

**Comment by guits** [comment]
jenkins test centos-container-purge

**Comment by guits** [comment]
jenkins test centos-container-podman

**Comment by guits** [comment]
jenkins test centos-container-all_daemons

**Comment by dsavineau** [comment]
jenkins test centos-container-podman

**Comment by dsavineau** [comment]
sounds good :+1:

**Pull request merged**